src/fsharp/service/ServiceParsedInputOps.fs

```diff
@@ -957,6 +960,9 @@ module ParsedInput =
                                 | _ -> 
                                     defaultTraverse expr
 
+                            // Unchecked.defaultof<str$>
+                            | SynExpr.TypeApp (typeArgsRange = range) when rangeContainsPos range pos ->
+                                Some CompletionContext.TypeApp
```

Comment:
In light of points 3 and 4, I'd rather remove `TypeApp` and abuse `CompletionContext.PatternType`. The reason is that we might like to have the `TypeApp` case carry further information in the future, and we'd be making a breaking change then.